bitswap/network/router.go

```diff
@@ -0,0 +1,136 @@
+package network
+
+import (
+	"context"
+	"time"
+
+	bsmsg "github.com/ipfs/boxo/bitswap/message"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	"go.uber.org/multierr"
+)
+
+type router struct {
+	Bitswap   BitSwapNetwork
+	HTTP      BitSwapNetwork
+	Peerstore peerstore.Peerstore
+}
+
+// New returns a BitSwapNetwork supported by underlying IPFS host.
+func New(pstore peerstore.Peerstore, bitswap BitSwapNetwork, http BitSwapNetwork) BitSwapNetwork {
+	if bitswap == nil && http == nil {
+		panic("bad exchange network router initialization: need bitswap or http")
+	}
+
+	if http == nil {
+		return bitswap
+	}
+
+	if bitswap == nil {
+		return http
+	}
+
+	return &router{
+		Peerstore: pstore,
+		Bitswap:   bitswap,
+		HTTP:      http,
+	}
+}
+
+func (rt *router) Start(receivers ...Receiver) {
+	rt.Bitswap.Start(receivers...)
+	rt.HTTP.Start(receivers...)
+}
+
+func (rt *router) Stop() {
+	rt.Bitswap.Stop()
+	rt.HTTP.Stop()
+}
+
+// Should be the same for both
+func (rt *router) Self() peer.ID {
+	return rt.Bitswap.Self()
+}
+
+func (rt *router) Ping(ctx context.Context, p peer.ID) ping.Result {
+	pi := rt.Peerstore.PeerInfo(p)
+	htaddrs, _ := SplitHTTPAddrs(pi)
+	if len(htaddrs.Addrs) > 0 {
+		return rt.HTTP.Ping(ctx, p)
+	}
+	return rt.Bitswap.Ping(ctx, p)
+}
+
+func (rt *router) Latency(p peer.ID) time.Duration {
+	pi := rt.Peerstore.PeerInfo(p)
+	htaddrs, _ := SplitHTTPAddrs(pi)
+	if len(htaddrs.Addrs) > 0 {
+		return rt.HTTP.Latency(p)
+	}
+	return rt.Bitswap.Latency(p)
+}
+
+func (rt *router) SendMessage(ctx context.Context, p peer.ID, msg bsmsg.BitSwapMessage) error {
+	// SendMessage is only used by bitswap server so we send a bitswap
+	// message.
+	return rt.Bitswap.SendMessage(ctx, p, msg)
+}
+
+// Connect attempts to connect to a peer. It prioritizes HTTP connections over
+// bitswap.
+func (rt *router) Connect(ctx context.Context, p peer.AddrInfo) error {
+	htaddrs, bsaddrs := SplitHTTPAddrs(p)
+	err := rt.HTTP.Connect(ctx, htaddrs)
+	if err != nil {
+		return rt.Bitswap.Connect(ctx, bsaddrs)
+	}
+	return nil
+}
+
+func (rt *router) DisconnectFrom(ctx context.Context, p peer.ID) error {
+	return multierr.Combine(
+		rt.HTTP.DisconnectFrom(ctx, p),
+		rt.Bitswap.DisconnectFrom(ctx, p),
+	)
```

Comment:
Give some thought here considering the note in `DisconnectFrom`.

Is there a change we are connected to the same peer over both Bitswap and HTTP? Oour intention is not to, but it is probably possible since `Connect` depends on provider records that can be anything.

Who calls DisconnectFrom, on what cases and what are the expectations?

Reply:
DisconnectFrom is only called internally from httpnet, and otherwise from bitswap Server when a failure receiving occurred.